src/rust/src/feature_writer.rs

```diff
@@ -60,19 +71,90 @@ impl<'a> FeatureWriter<'a> {
             t: Vec::new(),
             tm: Vec::new(),
             type_: GeometryType::Unknown,
+            dataset_type,
+            detect_type,
+            promote_to_multi,
             parts: Vec::new(),
             geom_state: GeomState::Normal,
             properties: Vec::new(),
             fbb: flatbuffers::FlatBufferBuilder::new(),
             bbox: NodeItem::create(0),
         }
     }
+    fn set_type(&mut self, geometry_type: GeometryType) -> Result<()> {
+        if geometry_type != self.dataset_type {
+            match self.dataset_type {
+                GeometryType::Unknown => {
+                    let type_ = match geometry_type {
+                        GeometryType::LineString if self.promote_to_multi => {
+                            self.geom_state = GeomState::ForceMulti;
+                            GeometryType::MultiLineString
+                        }
+                        GeometryType::Polygon if self.promote_to_multi => {
+                            self.geom_state = GeomState::ForceMulti;
+                            GeometryType::MultiPolygon
+                        }
+                        _ => geometry_type,
```

Comment:
I see the promotions for MultiLineString and MultiPolygon - do we need a similar promotion for MultiPoint?

Reply:
Probably yes. It's such a rare case that I forgot about.